src/gretel_trainer/relational/json.py

```diff
@@ -346,19 +364,24 @@ def _denormalize_json(
                 }
                 for k, v in flat_df.to_dict("index").items()
             }
+            # ...and after unflattening, go back to a dataframe
             dict_df = nulls_to_empty_dicts(
                 pd.DataFrame.from_dict(
                     {k: unflatten(v) for k, v in flat_dict.items()}, orient="index"
                 )
             )
-            nested_df = table_df.join(dict_df).drop(columns=col_names)
+
+            # Join the flattened dict_df onto table_df.
+            # You might think we could drop the original non-nested columns at this point,
+            # but under certain JSON shapes (e.g. nested lists of objects) we still need them present.
+            nested_df = table_df.join(dict_df)
```

Comment:
This is the only change we really had to make to fix the bug. Github is mucking up the diff a bit—compare it to the deleted line 354: we were dropping columns too aggressively, leading to `KeyError`s when performing `groupby` calls a few lines lower.